Changes to enable per channel support on dynamic linear.

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37829 Some TODO fixes.
* #37626 Perf optimization for conv and gemm kernels.
* #37624 Added per channel separate test cases for fc and deconv tests.
* **#37623 Changes to enable per channel support on dynamic linear.**
* #37622 Enabled per channel quantized static linear/conv
* #37621 Added per channel kernels for depthwise conv.
* #37620 Changes to enable per channel requant.
* #37619 Enable per channel zero point.
* #37618 Interface changes to enable per channel quant.

Summary:
Follows the same strategy as static linear. Supply pointer to an array of zero points and requantization scales. Thus same kernel now supports both per-channel and per-tensor linear. 

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

kimishpatel committed:
Changes to enable per channel support on dynamic linear.

Summary:
Follows the same strategy as static linear.
Same kernel now supports both per-channel and per-tensor linear.
Fixed fully connected test.

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear.
Same kernel now supports both per-channel and per-tensor linear.
Fixed fully connected test.

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear.
Same kernel now supports both per-channel and per-tensor linear.
Fixed fully connected test.

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear.
Same kernel now supports both per-channel and per-tensor linear.
Fixed fully connected test.

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear.
Same kernel now supports both per-channel and per-tensor linear.
Fixed fully connected test.

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear.
Same kernel now supports both per-channel and per-tensor linear.
Fixed fully connected test.

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear.
Same kernel now supports both per-channel and per-tensor linear.
Fixed fully connected test.

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear.
Same kernel now supports both per-channel and per-tensor linear.
Fixed fully connected test.

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear.
Same kernel now supports both per-channel and per-tensor linear.
Fixed fully connected test.

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear.
Same kernel now supports both per-channel and per-tensor linear.
Fixed fully connected test.

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear.
Same kernel now supports both per-channel and per-tensor linear.
Fixed fully connected test.

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear.
Same kernel now supports both per-channel and per-tensor linear.
Fixed fully connected test.

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear. Supply pointer to an array of zero points and requantization scales. Thus same kernel now supports both per-channel and per-tensor linear. 

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear. Supply pointer to an array of zero points and requantization scales. Thus same kernel now supports both per-channel and per-tensor linear. 

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear. Supply pointer to an array of zero points and requantization scales. Thus same kernel now supports both per-channel and per-tensor linear. 

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear. Supply pointer to an array of zero points and requantization scales. Thus same kernel now supports both per-channel and per-tensor linear. 

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear. Supply pointer to an array of zero points and requantization scales. Thus same kernel now supports both per-channel and per-tensor linear. 

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

dreiss commented on aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:
```diff
@@ -256,13 +250,22 @@ at::Tensor PackedLinearWeightsQnnp::apply_dynamic_impl(at::Tensor input) {
+    requantization_scale =
+        generate_requantization_scales(w_scales, q_params.scale, 1.f);
+
+    // TODO Kimish, we are allocating affine_quantized regardless of per channel or not.
+    // This allocation is actually used only for packing weight and thus will be freed.
+    // Still we should be consistent. Fix this.
```
Are the scale and offset even used here?  Could this just be a normal CPU tensor of dtype uint8?

kimishpatel replied:
Yes it can be. Let me do that after this stack lands.

dreiss reviewed (approved):
Didn't review the microkernel changes, but the rest looks good.

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear. Supply pointer to an array of zero points and requantization scales. Thus same kernel now supports both per-channel and per-tensor linear. 

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear. Supply pointer to an array of zero points and requantization scales. Thus same kernel now supports both per-channel and per-tensor linear. 

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear. Supply pointer to an array of zero points and requantization scales. Thus same kernel now supports both per-channel and per-tensor linear. 

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear. Supply pointer to an array of zero points and requantization scales. Thus same kernel now supports both per-channel and per-tensor linear. 

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear. Supply pointer to an array of zero points and requantization scales. Thus same kernel now supports both per-channel and per-tensor linear. 

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear. Supply pointer to an array of zero points and requantization scales. Thus same kernel now supports both per-channel and per-tensor linear. 

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear. Supply pointer to an array of zero points and requantization scales. Thus same kernel now supports both per-channel and per-tensor linear. 

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel support on dynamic linear."


Summary:
Follows the same strategy as static linear. Supply pointer to an array of zero points and requantization scales. Thus same kernel now supports both per-channel and per-tensor linear. 

Test Plan:
qnnpack tests
q8gemm
fully-connected-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339040](https://our.internmc.facebook.com/intern/diff/D21339040)

[ghstack-poisoned]
